# AnCovMulti
Ancestral Covid Variant Analysis

The current program v0.1 created Dec. 28, 2020

AnCovMulti is written in Golang to analyze variants in Covid amplicon sequence data. The first version was written December 28, 2020 by David Pollock with consultation from Richard Goldstein. It is currently restricted to Illumina reads. 
Please cite Kemp et al., 2020.

Further information about the program and how to run it can be found in the file AnCovMulti_Readme3.docx

All code and associated reference stuff is included in the compressed file AnCovMulti Demo Small.zip
The compressed .fq sequence file is included in a separate file and should be deposited in the folder named "sequences" when the main code folder is unzipped.

The behaviors if you deviate from the prescribed file and folder formats are unknown

Neutralising antibodies in Spike mediated SARS-CoV-2 adaptation, 
Kemp SA, Collier DA, Datir R, Ferreira IATM , Gayed S, Jahun A, Hosmillo M, Rees-Spear C, Mlcochova P, Ines Ushiro Lumb, David J Roberts, Anita Chandra, Temperton N, The CITIID-NIHR BioResource COVID-19 Collaboration, The COVID-19 Genomics UK (COG-UK) Consortium, Sharrocks K, Blane E, Briggs JAG, van Gils MJ11, Smith KGC, Bradley JR, Smith C, Doffinger R, Ceron-Gutierrez L, Barcenas-Morales G, Pollock DD, Goldstein RA, Smielewska A, Skittrall JP, Gouliouris T, Goodfellow IG, Gkrania-Klotsas E, Illingworth CJR, McCoy LE, Gupta RK 




